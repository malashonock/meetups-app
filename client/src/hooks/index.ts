export * from './useMeetupQuery';
export * from './useNewsArticleQuery';
export * from './useStaticFileQuery';
export * from './useAuthStore';
