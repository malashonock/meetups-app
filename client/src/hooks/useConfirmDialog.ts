import { useContext } from 'react';

import { ConfirmDialogContext, ConfirmFn } from 'components';
import { Optional } from 'types';

export const useConfirmDialog = (): ConfirmFn => {
  return useContext(ConfirmDialogContext);
};
