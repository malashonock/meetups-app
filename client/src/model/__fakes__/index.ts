export * from './user.fake';
export * from './meetup.fake';
export * from './file.fake';
