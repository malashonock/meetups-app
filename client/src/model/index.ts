export * from './meetup.model';
export * from './news.model';
export * from './user.model';
export * from './login.model';
