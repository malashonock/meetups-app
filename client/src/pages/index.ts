export * from './LoginPage/LoginPage';
export * from './MeetupsPage/MeetupsPage';
export * from './ViewMeetupPage/ViewMeetupPage';
export * from './CreateMeetupPage/CreateMeetupPage';
export * from './EditMeetupPage/EditMeetupPage';
export * from './NewsPage/NewsPage';
export * from './ViewNewsPage/ViewNewsPage';
export * from './CreateNewsPage/CreateNewsPage';
export * from './EditNewsPage/EditNewsPage';
export * from './NotFoundPage/NotFoundPage';
