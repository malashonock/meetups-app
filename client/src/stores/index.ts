export * from './root.store';
export * from './auth.store';
export * from './user.store';
export * from './ui.store';
export * from './meetup.store';
export * from './news.store';
