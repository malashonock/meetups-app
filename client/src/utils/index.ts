export * from './string';
export * from './date';
export * from './number';
export * from './file';
export * from './array';
export * from './test';
export * from './dom';
