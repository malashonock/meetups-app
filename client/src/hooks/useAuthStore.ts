import { useContext } from 'react';

import { AuthStore } from 'stores';
import { RootContext } from 'components';

export const useAuthStore = (): AuthStore => {
  return useContext(RootContext).authStore;
};
