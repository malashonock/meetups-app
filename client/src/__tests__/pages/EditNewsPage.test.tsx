/* eslint-disable testing-library/no-unnecessary-act */

import { PropsWithChildren } from 'react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { act, fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { EditNewsPage } from 'pages';
import { NewsFields } from 'model';
import { mockImageWithUrl, mockNewsArticle } from 'model/__fakes__';
import { dropFile } from 'utils';
import { useLocale, useNewsArticle } from 'hooks';
import { Locale, News } from 'stores';

jest.mock('utils/file');

// Mock hooks
jest.mock('hooks', () => {
  return {
    ...jest.requireActual('hooks'),
    useNewsArticle: jest.fn(),
    useLocale: jest.fn(),
  };
});
const mockUseNewsArticle = useNewsArticle as jest.MockedFunction<
  typeof useNewsArticle
>;
const mockUseLocale = useLocale as jest.MockedFunction<typeof useLocale>;
const mockUpdatedNewsArticleUpdate = jest.spyOn(News.prototype, 'update');

beforeEach(() => {
  mockUseNewsArticle.mockReturnValue(mockNewsArticle);
  mockUseLocale.mockReturnValue([Locale.RU, jest.fn()]);
});

afterEach(() => {
  jest.resetAllMocks();
});

const MockRouter = ({ children }: PropsWithChildren): JSX.Element => (
  <MemoryRouter initialEntries={['/news/aaa', '/news/aaa/edit']}>
    <Routes>
      <Route path="/news">
        <Route index element={<h1>News page</h1>} />
        <Route path=":id">
          <Route index element={<h1>View news article</h1>} />
          <Route path="edit" element={children} />
        </Route>
      </Route>
    </Routes>
  </MemoryRouter>
);

const getTitleInput = () =>
  screen.getByLabelText('formFields.news.title.label');
const getTextInput = () => screen.getByLabelText('formFields.news.text.label');
const getImageDropbox = () => screen.getByTestId('image-dropbox');
const getImagePreview = () => screen.getByTestId('image-preview');
const getImage = () =>
  screen.getByAltText('imagePreview.imgAlt') as HTMLImageElement;
const getSubmitBtn = () => screen.getByText('formButtons.save');
const getCancelBtn = () => screen.getByText('formButtons.cancel');
const getClearImageBtn = () => screen.getByTestId('clear-button');

const mockUpdatedNews: NewsFields = {
  title: 'Updated news title',
  text: 'Updated news text',
  image: mockImageWithUrl,
};

describe('EditNewsPage', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<EditNewsPage />, { wrapper: MockRouter });

    expect(asFragment()).toMatchSnapshot();
  });

  it('on open, should pre-fill form inputs with the current field values', () => {
    render(<EditNewsPage />, { wrapper: MockRouter });

    expect(getTitleInput()).toHaveValue(mockNewsArticle.title);
    expect(getTextInput()).toHaveValue(mockNewsArticle.text);
    expect(getImage().src).toBe(mockNewsArticle.image?.url);
  });

  it('should accept user input', async () => {
    render(<EditNewsPage />, { wrapper: MockRouter });

    await act(() => {
      userEvent.clear(getTitleInput());
      userEvent.type(getTitleInput(), mockUpdatedNews.title);
    });
    expect(getTitleInput()).toHaveValue(mockUpdatedNews.title);

    await act(() => {
      userEvent.clear(getTextInput());
      userEvent.type(getTextInput(), mockUpdatedNews.text);
    });
    expect(getTextInput()).toHaveValue(mockUpdatedNews.text);

    await act(() => {
      userEvent.click(getClearImageBtn());
    });
    expect(getImageDropbox()).toBeInTheDocument();

    await act(() => {
      dropFile(getImageDropbox(), mockImageWithUrl);
    });
    expect(getImagePreview()).toBeInTheDocument();
    expect(getImage().src).toBe(mockImageWithUrl.url);
  });

  it('should validate form fields', async () => {
    render(<EditNewsPage />, { wrapper: MockRouter });

    expect(getSubmitBtn()).toBeDisabled();

    await act(() => {
      userEvent.clear(getTitleInput());
      fireEvent.blur(getTitleInput());
    });
    expect(
      screen.getByText('formFields.news.title.errorText'),
    ).toBeInTheDocument();

    await act(() => {
      userEvent.clear(getTextInput());
      fireEvent.blur(getTextInput());
    });
    expect(
      screen.getByText('formFields.news.text.errorText'),
    ).toBeInTheDocument();
  });

  it('should handle form submit', async () => {
    render(<EditNewsPage />, { wrapper: MockRouter });

    await act(() => {
      userEvent.clear(getTitleInput());
      userEvent.type(getTitleInput(), mockUpdatedNews.title);
    });

    await act(() => {
      userEvent.clear(getTextInput());
      userEvent.type(getTextInput(), mockUpdatedNews.text);
    });

    await act(() => {
      userEvent.click(getClearImageBtn());
    });

    await act(() => {
      dropFile(getImageDropbox(), mockImageWithUrl);
    });

    await act(() => {
      userEvent.click(getSubmitBtn());
    });

    expect(mockUpdatedNewsArticleUpdate).toHaveBeenCalledWith(mockUpdatedNews);
    expect(screen.getByText('News page')).toBeInTheDocument();
  });

  it('should be able to navigate back to news page', async () => {
    render(<EditNewsPage />, { wrapper: MockRouter });

    await act(() => {
      userEvent.click(getCancelBtn());
    });

    expect(screen.getByText('View news article')).toBeInTheDocument();
  });

  it('should render a Loading spinner if news article is undefined', () => {
    mockUseNewsArticle.mockReturnValue(undefined);
    render(<EditNewsPage />, { wrapper: MockRouter });
    expect(screen.getByText('loadingText.newsArticle')).toBeInTheDocument();
  });

  it('should render a Loading spinner while news article is loading', () => {
    const newsArticle = new News(mockNewsArticle);
    newsArticle.isLoading = true;
    mockUseNewsArticle.mockReturnValue(newsArticle);
    render(<EditNewsPage />, { wrapper: MockRouter });
    expect(screen.getByText('loadingText.newsArticle')).toBeInTheDocument();
  });

  it('should render Not Found page if an error occurred while loading the news article', () => {
    const newsArticle = new News(mockNewsArticle);
    newsArticle.isError = true;
    mockUseNewsArticle.mockReturnValue(newsArticle);
    render(<EditNewsPage />, { wrapper: MockRouter });
    expect(screen.getByText('notFoundPage.title')).toBeInTheDocument();
  });
});
