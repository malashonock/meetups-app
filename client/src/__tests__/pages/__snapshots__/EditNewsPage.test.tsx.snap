// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditNewsPage should match snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <section
      class="container"
    >
      <h1
        class="typography h1 heading"
      >
        editNewsPage.title
      </h1>
      <div
        class="contentWrapper"
      >
        <div
          class="textSection main"
        >
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="image"
            >
              formFields.news.image.label
            </label>
            <div
              class="preview large"
              data-testid="image-preview"
            >
              <figure
                class="image"
              >
                <img
                  alt="imagePreview.imgAlt"
                  src="http://localhost/test.jpg"
                />
              </figure>
              <button
                class="clearBtn"
                data-testid="clear-button"
                type="button"
              >
                <svg>
                  change-photo.svg
                </svg>
              </button>
            </div>
          </div>
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="title"
            >
              formFields.news.title.label
            </label>
            <input
              class="input default"
              data-testid="text-input"
              id="title"
              name="title"
              type="text"
              value="Test news title"
            />
          </div>
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="text"
            >
              formFields.news.text.label
            </label>
            <div
              class="container"
            >
              <textarea
                class="input default textArea"
                data-testid="text-area"
                id="text"
                name="text"
                rows="1"
                style="height: 2px;"
              >
                Test news text
              </textarea>
            </div>
          </div>
        </div>
        <div
          class="textSection actions"
        >
          <button
            class="actionButton button default"
            id="btn-cancel"
            type="button"
          >
            formButtons.cancel
          </button>
          <button
            class="actionButton button primary"
            disabled=""
            id="btn-save"
            type="submit"
          >
            formButtons.save
          </button>
        </div>
      </div>
    </section>
  </form>
</DocumentFragment>
`;
